src/target/riscv/riscv.h

```diff
@@ -193,7 +193,19 @@ struct riscv_info {
 	 * was resumed. */
 	int (*resume_go)(struct target *target);
 	int (*step_current_hart)(struct target *target);
-	int (*on_halt)(struct target *target);
+
+	/* These get called from riscv_poll_hart(), which is a house of cards
+	 * together with openocd_poll(), so be careful not to upset things too
+	 * much. */
+	int (*became_halted)(struct target *target);
+	int (*became_running)(struct target *target);
+	int (*became_unavailable)(struct target *target);
+	int (*became_nonexistent)(struct target *target);
```

Comment:
I wonder whether handling of "became_nonexistent" is needed. The RISC-V debug spec says: `"Harts are nonexistent if they will never be part of this hardware platform, no matter how long a user waits"`.

So in a compliant hardware, nonexistent harts should remain nonexistent forever, and should never change state.

Did you mean to handle this "transition to nonexistent state" just in case; just for additional safety?

Reply:
You are right in that technically this should never occur. I'll add a comment and a warning when this does happen.